Update auto_functionalize schema

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118810
* __->__ #118809

- Moved the dictionary arguments to the node's kwargs as dicts are not
  valid inputs.
- Inlined the mutated arguments to the output. Originally, the output of
  auto_functionalize was the operator output
  and a list of mutated arguments (ex. [op_out1, op_out2, [mutated_arg1,
  mutated_arg2]]. However this is not easily exportable. Now, it will
  just be [op_out1, op_out2, mutated_arg1, mutated_arg2].

Differential Revision: [D53331040](https://our.internmc.facebook.com/intern/diff/D53331040)

angelayi committed:
Update auto_functionalize schema

- Moved the dictionary arguments to the node's kwargs as dicts are not
  valid inputs.
- Inlined the mutated arguments to the output. Originally, the output of
  auto_functionalize was the operator output
  and a list of mutated arguments (ex. [op_out1, op_out2, [mutated_arg1,
  mutated_arg2]]. However this is not easily exportable. Now, it will
  just be [op_out1, op_out2, mutated_arg1, mutated_arg2].

[ghstack-poisoned]

suo commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -54,12 +54,12 @@ def __call__(
         op: torch._ops.OpOverload,
         mutated_args_names: List[str],
-        kwargs: Dict[str, Any],
+        **kwargs: Dict[str, Any],
```
this is a bit weird in the case that the operator has a kwargs named "op" or "mutated_arg_names".

```
>>> def foo(a, **kwargs):
...     print(a, kwargs)
... 
>>> foo(10, a=5)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: foo() got multiple values for argument 'a'
```

At least we get a loud error in this case, so I'm okay with it in principle. Maybe rename `op` to something less common and prefix both arguments with `_` to reduce the chance of collision.

This comment applies to every function signature where we did this change.

suo reviewed (commented):
lgtm, but let's let @zou3519 sign off

angelayi committed:
Update on "Update auto_functionalize schema"


- Moved the dictionary arguments to the node's kwargs as dicts are not
  valid inputs.
- Inlined the mutated arguments to the output. Originally, the output of
  auto_functionalize was the operator output
  and a list of mutated arguments (ex. [op_out1, op_out2, [mutated_arg1,
  mutated_arg2]]. However this is not easily exportable. Now, it will
  just be [op_out1, op_out2, mutated_arg1, mutated_arg2].

[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -52,14 +52,14 @@ def __init__(self):
-        op: torch._ops.OpOverload,
-        mutated_args_names: List[str],
-        kwargs: Dict[str, Any],
+        _nonfunctional_op: torch._ops.OpOverload,
+        _mutated_args_names: List[str],
```
Why are these underscore?

zou3519 replied:
Oh I see

zou3519 commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -34,7 +34,7 @@
 
 class AutoFunctionalized(HigherOrderOperator):
-    """auto_functionalized(op, mutated_args_names, kwargs)
+    """auto_functionalized(_nonfunctional_op, mutated_args_names, kwargs)
```
_mutated_args_names

zou3519 replied:
And `**kwargs`

zou3519 replied:
Can you leave a comment somewhere saying the reason why these are underscore? (to prevent name collisions with **kwargs)

zou3519 commented on torch/_higher_order_ops/auto_functionalize.py:
```diff
@@ -52,14 +52,14 @@ def __init__(self):
     ) -> Tuple[Any, Tuple[Tensor, ...]]:
-        assert can_auto_functionalize(op)
-        assert isinstance(mutated_args_names, list)
+        assert can_auto_functionalize(_nonfunctional_op)
```
nit: prefer `_mutable_op` over `_nonfunctional_op`

zou3519 reviewed (approved):
LGTM, see comments

angelayi committed:
Update on "Update auto_functionalize schema"


- Moved the dictionary arguments to the node's kwargs as dicts are not
  valid inputs.
- Inlined the mutated arguments to the output. Originally, the output of
  auto_functionalize was the operator output
  and a list of mutated arguments (ex. [op_out1, op_out2, [mutated_arg1,
  mutated_arg2]]. However this is not easily exportable. Now, it will
  just be [op_out1, op_out2, mutated_arg1, mutated_arg2].

[ghstack-poisoned]

angelayi committed:
Update on "Update auto_functionalize schema"


- Moved the dictionary arguments to the node's kwargs as dicts are not
  valid inputs.
- Inlined the mutated arguments to the output. Originally, the output of
  auto_functionalize was the operator output
  and a list of mutated arguments (ex. [op_out1, op_out2, [mutated_arg1,
  mutated_arg2]]. However this is not easily exportable. Now, it will
  just be [op_out1, op_out2, mutated_arg1, mutated_arg2].

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D53331040).
